general/communication.hpp

```diff
@@ -19,31 +19,30 @@
 #include "array.hpp"
 #include "table.hpp"
 #include "sets.hpp"
+#include "globals.hpp"
 #include <mpi.h>
 
+
 namespace mfem
 {
 
 /** @brief A simple convenience class that calls MPI_Init() at construction and
-    MPI_Finalize() at destruction. It also provides easy access to
-    MPI_COMM_WORLD's rank and size. */
+    MPI_Finalize() at destruction. It also provides easy access to the rank and
+    size of the given "world" communicator. */
 class MPI_Session
 {
 protected:
    int world_rank, world_size;
-   void GetRankAndSize()
-   {
-      MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
-      MPI_Comm_size(MPI_COMM_WORLD, &world_size);
-   }
+   void GetRankAndSize(MPI_Comm);
 public:
-   MPI_Session() { MPI_Init(NULL, NULL); GetRankAndSize(); }
-   MPI_Session(int &argc, char **&argv)
-   { MPI_Init(&argc, &argv); GetRankAndSize(); }
+   MPI_Session(MPI_Comm world_comm = MPI_COMM_WORLD)
```

Comment:
Looking at this again, I don't think any communicator other than `MPI_COMM_WORLD` makes sense in this class - it is meant to initialize MPI, so you cannot have any other communicators.

We should probably just remove this optional parameter.

Reply:
Agreed